314 - Implement ids-search-field

vivz753 opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**

Adds the web component Ids Search Field

**Related github/jira issue (required)**:
- Closes #314 

**Steps necessary to review your pull request (required)**:
- check out http://localhost:4300/ids-header
- check out http://localhost:4300/ids-search-field

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

tmcconechy commented on demos/ids-search-field/index.yaml:
```diff
@@ -0,0 +1,9 @@
+ component: Search Field 
+ link: ids-search-field
+ description: Interactive input for typing in to search for results
+ category: Input
```
```suggestion
 category: Form Inputs
```

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,221 @@
+      ...super.attributes,
+      attributes.VALUE,
+      attributes.PLACEHOLDER,
+      // attributes.ID,
```
Should remove all the commented out code?

vivz753 replied:
Done

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,221 @@
+    return this.getAttribute(attributes.LABEL) || DEFAULT_LABEL;
+  }
+
+  set disabled(value) {
```
Add jsdoc to all methods/setters/getters?

vivz753 replied:
Done

tmcconechy commented on src/components/ids-trigger-field/ids-trigger-field.scss:
```diff
@@ -95,6 +95,11 @@ $input-size-default: map-get($input-sizes, 'md');
     @include text-slate-30();
```
Not sure about these changes? Are they all correct. I made corrections on my dropdown PR to make the readonly states look correct and these seem conflicting. Can you try to use this for ids-trigger-field.scss and see if it resolves your issue? Or maybe we skip this and add it after my changes (as a new issue?)

```
@import '../../core/ids-base';
@import '../ids-input/ids-input';

.ids-trigger-field {
  @include flex();
  @include mb-16();

  align-items: flex-start;
  flex-direction: column;
}

:host(ids-trigger-field[disabled]) {
  pointer-events: none;
}

.ids-trigger-field-content {
  @include antialiased();
  @include bg-white();
  @include box-border();
  @include border-1();
  @include border-slate-40();
  @include border-solid();
  @include font-sans();
  @include flex();
  @include rounded-default();
  @include text-16();
  @include text-black();

  align-items: center;
  -webkit-appearance: none;
  border-collapse: separate;
  border-radius: 2px;
  height: $input-field-height-md;
  width: 100%;

  &:hover {
    @include border-slate-90();
  }

  &.disabled {
    @include border-slate-30();
    @include text-slate-30();
  }

  &[readonly] {
    @include bg-slate-20();
    @include border-slate-40();
  }

  // input message: [alert, error, info, success]
  &.alert,
  &.alert:hover {
    @include border-alert-warning();
    @include text-alert-warning();

    &:focus-within {
      @include border-alert-warning();

      box-shadow: rgba(249, 131, 0, 0.3) 0 0 3px 2px;
    }

    &[disabled] {
      @include border-amber-30();
      @include text-amber-30();
    }
  }

  &.error,
  &.error:hover {
    @include border-alert-danger();
    @include text-alert-danger();

    &:focus-within {
      @include border-alert-danger();
      @include shadow-error();
    }

    &[disabled] {
      @include border-ruby-30();
      @include text-ruby-30();
    }
  }

  &.info,
  &.info:hover {
    @include border-alert-info();
    @include text-alert-info();

    &:focus-within {
      @include border-alert-info();

      box-shadow: rgba(0, 102, 212, 0.3) 0 0 3px 2px;
    }

    &[disabled] {
      @include border-azure-30();
      @include text-azure-30();
    }
  }

  &.success,
  &.success:hover {
    @include border-alert-success();
    @include text-alert-success();

    &:focus {
      @include border-alert-success();

      box-shadow: rgba(42, 195, 113, 0.3) 0 0 3px 2px;
    }

    &[disabled] {
      @include border-emerald-30();
      @include text-emerald-30();
    }
  }
}

// input sizes: [xs, sm, mm, md, lg, full]
.ids-trigger-field.xs .ids-trigger-field-content {
  max-width: $input-size-xs;
}

.ids-trigger-field.sm .ids-trigger-field-content {
  max-width: $input-size-sm;
}

.ids-trigger-field.mm .ids-trigger-field-content {
  max-width: $input-size-mm;
}

.ids-trigger-field.md .ids-trigger-field-content {
  max-width: $input-size-md;
}

.ids-trigger-field.lg .ids-trigger-field-content {
  max-width: $input-size-lg;
}

.ids-trigger-field.full .ids-trigger-field-content {
  max-width: $input-size-full;
}

.ids-trigger-field:focus-within,
.ids-trigger-field.is-active {
  .ids-trigger-field-content:not(.error):not(.info):not(.success) {
    @include border-azure-60();
    @include shadow();

    outline: none;
    outline-color: transparent;
  }
}

::slotted(ids-trigger-button) {
  @include relative();

  flex: 0;
  z-index: 2;
}

::slotted(ids-input) {
  @include relative();

  flex: 1;
  max-width: $input-size-default;
  z-index: 1;
}

// Handle Themes
.ids-trigger-field[mode='dark'] .ids-trigger-field-content {
  @include bg-transparent();
  @include border-slate-40();
  @include text-white();

  &.disabled {
    @include border-slate-60();
    @include text-slate-60();
  }

  &[readonly] {
    @include bg-slate-70();
    @include border-slate-40();
    @include text-slate-10();
  }

  &.bg-transparent[readonly] {
    @include bg-transparent();
  }

  &:focus-within {
    @include border-azure-60();
  }
}

.ids-trigger-field[mode='contrast'] .ids-trigger-field-content {
  @include bg-white();
  @include border-slate-40();
  @include text-black();

  &.disabled {
    @include border-slate-30();
    @include text-slate-30();
  }

  &[readonly] {
    @include bg-slate-20();
    @include border-slate-40();
    @include text-slate-100();
  }

  &.bg-transparent[readonly] {
    @include bg-white();
  }

  &:focus-within {
    @include border-azure-90();
  }
}

.ids-trigger-field[version='classic'][mode='light'] .ids-trigger-field-content {
  @include bg-transparent();
  @include border-graphite-40();
  @include text-graphite-100();

  &.disabled {
    @include border-graphite-30();
    @include text-graphite-40();
  }

  &[readonly] {
    @include bg-graphite-20();
    @include border-graphite-40();
    @include text-classic-slate-100();
  }

  &.bg-transparent[readonly] {
    @include bg-transparent();
  }

  &:focus-within {
    @include border-azure-60();
  }
}

.ids-trigger-field[version='classic'][mode='dark'] .ids-trigger-field-content {
  @include bg-transparent();
  @include border-slate-40();
  @include text-white();

  &.disabled {
    @include border-classic-slate-40();
    @include text-classic-slate-50();
  }

  &[readonly] {
    @include bg-classic-slate-60();
    @include border-classic-slate-40();
    @include text-white();
  }

  &.bg-transparent[readonly] {
    @include bg-transparent();
  }

  &:focus-within {
    @include border-azure-60();
  }
}

.ids-trigger-field[version='classic'][mode='contrast'] .ids-trigger-field-content {
  @include bg-white();
  @include border-graphite-80();
  @include text-graphite-100();

  &.disabled {
    @include bg-transparent();
    @include border-slate-50();
    @include text-graphite-80();
  }

  &[readonly] {
    @include bg-slate-30();
    @include border-slate-80();
    @include text-graphite-100();
  }

  &.bg-transparent[readonly] {
    @include bg-white();
  }

  &:focus-within,
  &:focus[readonly] {
    @include border-azure-60();
  }
}

// Not sure why we need to do this
.validation-message {
  margin-bottom: 12px;
  margin-top: -12px;
}

```

vivz753 replied:
As discussed, going to wait for PR's addressing layout/styles in ids-trigger-field and ids-input to be merged in with main first, then will address the merge conflicts on my branch

tmcconechy reviewed (commented):
Some initial comments. Also for the example

1. https://main-enterprise.demo.design.infor.com/components/searchfield/example-index.html 
2. Instead of Pukichu -> use "boring" business terms or generic words like "search"
3. The search field does have an arrow -> so that should be removed

vivz753 commented:
> Some initial comments. Also for the example
> 
> 1. Instead of Pukichu -> use "boring" business terms or generic words like "search"
> 2. The search field does have an arrow -> so that should be removed see https://main-enterprise.demo.design.infor.com/components/searchfield/example-index.html
> 3. What about the in the header search? https://main-enterprise.demo.design.infor.com/components/header/example-searchfield-expanded.html

Addressed all of the above

vivz753 commented on src/components/ids-input/ids-input.scss:
```diff
@@ -335,7 +335,7 @@ $input-compact-height: $input-field-height-xs;
     @include text-slate-10();
   }
 
-  &.bg-transparent[readonly] {
```
There is no reason or logic in the code that suggests that bg-transparent and readonly status should be coupled--separated the CSS classes from the readonly attribute

tmcconechy replied:
Hmm i wouldnt remove this. Without testing it on dropdown. The use case is sometimes the input should be readonly state but not look readonly one is the dropdown where you dont edit the text you select it from a list but it still looks enabled.

Should we leave this for now?

vivz753 replied:
Thanks for the context. We can always change it to work for future requirements, but i'm actually using the bg-transparent class for the ids-input component inside of ids-trigger-field for my component. That way I can just set one color for ids-trigger-field and not have to change ids-input--needed this for the style on the header search bar. 

I checked twice and there is no code in the repo besides the ids-input test that uses the bg-transparent attribute, so I'm going to say that it shouldn't affect dropdown. Unless you have a PR that uses it?

tmcconechy replied:
Yes i use it. So it will break me or i will have to add it back

vivz753 replied:
https://github.com/infor-design/enterprise-wc/compare/132-dropdown#diff-351a0d15e02b8c425f71e5c8338b1fb762bfee79a44261c3e9e8aeb1d9a91309R128

Your code will never have readonly and bg-transparent as true based on your logic, so this change shouldn't affect you at all actually.

vivz753 commented on src/components/ids-radio/ids-radio-group.js:
```diff
@@ -50,7 +50,8 @@ class IdsRadioGroup extends mix(IdsElement).with(
    */
   static get attributes() {
     return [
-      ...attributes.DISABLED,
```
Fixed incorrect use of spread operator

vivz753 commented on src/components/ids-search-field/ids-search-field.scss:
```diff
@@ -0,0 +1,35 @@
+    @include ml-8();
+  }
+
+  &.color-variant-alternate {
```
alternate style--for when using inside of ids-header

vivz753 commented on src/components/ids-trigger-field/ids-trigger-field.js:
```diff
@@ -93,15 +95,17 @@ class IdsTriggerField extends IdsInput {
       <div class="ids-trigger-field ${this.size}" part="field">
         ${ this.label !== '' ? `<label
-          ${ buildClassAttrib('ids-label-text', this.disabled !== null && 'disabled', this.validate !== null && 'required') }
+          ${ buildClassAttrib('ids-label-text', this.disabled && 'disabled', this.readonly && 'readonly') }
```
Add missing readonly status for ids-trigger-field

tmcconechy replied:
I did this on my dropdown PR so this could conflict. i also did it differently

vivz753 commented on src/components/ids-trigger-field/ids-trigger-field.js:
```diff
@@ -185,9 +190,11 @@ class IdsTriggerField extends IdsInput {
     this.setAttribute(attributes.TABBABLE, value.toString());
-    button.tabbable = isTabbable;
+    const button = this.querySelector('ids-trigger-button');
+    if (button) {
```
Error gets thrown if trigger button is left empty--add a check for it

tmcconechy replied:
This may also conflict. Heads up

vivz753 commented on src/mixins/ids-color-variant-mixin/README.md:
```diff
@@ -10,7 +10,7 @@ The usage of this mixin would be:
 1. Import the mixin and add it to the `mix` list
 1. Manually add the `color-variant` property to your example, or have your component set this property automatically.
-1. Add a property to your component called `availableColorVariants` that provides an array of strings containing names of color variants other than the default (In JS, the default variant is `null`).
+1. Add a property to your component called `colorVariants` that provides an array of strings containing names of color variants other than the default (In JS, the default variant is `null`).
```
Color Variant Mixin updated to use colorVariants instead of availableColorVariants, but was not updated in the README

vivz753 commented on src/mixins/ids-color-variant-mixin/ids-color-variant-mixin.js:
```diff
@@ -82,18 +82,19 @@ const IdsColorVariantMixin = (superclass) => class extends superclass {
-        cl.add(variantClass);
-      } else if (variantClass !== thisVariantClass && cl.contains(variantClass)) {
-        cl.remove(variantClass);
-      }
```
This above for loop runs O(n*c) times where n is the length of colorVariants[] and c is the length of classList[]--we can optimize it to O(c) where we just iterate through the classList to delete any other color-variants, then add the variantClass which is O(1)--see new algorithm below

tmcconechy replied:
not sure but cant hurt to try.

vivz753 commented on demos/ids-header/example.html:
```diff
@@ -11,6 +11,9 @@
     </ids-toolbar-section>
+    <ids-toolbar-section type="search" align="center">
+      <ids-search-field color-variant="alternate" label=""></ids-search-field>
+    </ids-toolbar-section>
```
Add search-field example to ids-header demo

vivz753 commented on demos/ids-trigger-field/example.html:
```diff
@@ -105,4 +105,24 @@
+        <ids-icon slot="icon" icon="search"></ids-icon>
+      </ids-trigger-button>
+    </ids-trigger-field>
+  </ids-layout-grid-cell>
```
Add readonly example to ids-trigger-field demo

vivz753 commented on src/components/ids-trigger-field/ids-trigger-field.scss:
```diff
@@ -95,6 +94,11 @@ $input-size-default: map-get($input-sizes, 'md');
+    @include bg-slate-20();
+    @include text-black();
+  }
+
```
Add missing readonly style to ids-trigger field--possible conflict with another PR, shouldn't be difficult to resolve

vivz753 commented on src/components/ids-trigger-field/ids-trigger-field.scss:
```diff
@@ -200,13 +204,13 @@ $input-size-default: map-get($input-sizes, 'md');
   }
 
-  &[readonly] {
+  &.readonly {
```
There is a discrepancy between using readonly as an attribute or class for the css selector... Most other components (i.e. ids-input) follow a pattern which uses states (i.e. disabled, readonly) as an attribute for css selector. Adding it to the TODO to refactor this to follow that pattern.

tmcconechy replied:
Another possible conflict

tmcconechy commented:
WIP until dropdown is merged and synced

tmcconechy commented:
dropdown is merged so all the trigger changes are in. so this is ready to fix now 👍🏻

tmcconechy commented:
Please take off WIP if read to review. I see a couple things in the images:

1. something made one field really wide https://percy.io/00aaa841/enterprise-wc/builds/12890574/changed/728567092?browser=chrome&browser_ids=16&subcategories=unreviewed%2Cchanges_requested&viewLayout=side-by-side&viewMode=new&width=1280&widths=375%2C1280
2. For the header example i think we need one with it looking like https://main-enterprise.demo.design.infor.com/components/header/example-index.html (but with the search expanded only) (can add collapsible option too) but it should be on the left side not centered as per the images

vivz753 commented:
> Please take off WIP if read to review. I see a couple things in the images:
> 
> 1. something made one field really wide https://percy.io/00aaa841/enterprise-wc/builds/12890574/changed/728567092?browser=chrome&browser_ids=16&subcategories=unreviewed%2Cchanges_requested&viewLayout=side-by-side&viewMode=new&width=1280&widths=375%2C1280
> 2. For the header example i think we need one with it looking like https://main-enterprise.demo.design.infor.com/components/header/example-index.html (but with the search expanded only) (can add collapsible option too) but it should be on the left side not centered as per the images

1. Added a no-margins attribute for ids-trigger-field, which is toggled true for ids-search-field (screenshots for ids-trigger-field should remain the same) 

2. The compact header was a feature of phase 2, let's get this PR through first before working on that. e

vivz753 commented:
![Screen Shot 2021-09-24 at 11 05 33 AM](https://user-images.githubusercontent.com/22609861/134697666-daa20ab0-7dc7-4985-bbde-67e9028e546f.png)

Like this?

tmcconechy commented:
Can we make a totally new header example and not touch the main one. Like:

![Screen Shot 2021-09-24 at 11 07 18 AM](https://user-images.githubusercontent.com/814283/134697940-186a776c-c2f1-425d-b74c-4f0f2460e6ae.png)

But yours is maybe close just the icon color is wrong. Also keep in mind the color can change

vivz753 commented:
> Can we make a totally new header example and not touch the main one. Like:
> 
> ![Screen Shot 2021-09-24 at 11 07 18 AM](https://user-images.githubusercontent.com/814283/134697940-186a776c-c2f1-425d-b74c-4f0f2460e6ae.png)
> 
> But yours is maybe close just the icon color is wrong. Also keep in mind the color can change on the header

Done. See updated screenshot above

vivz753 commented:
New Header example w search field 

![Screen Shot 2021-09-24 at 11 05 33 AM](https://user-images.githubusercontent.com/22609861/134722672-c0df0555-2554-4538-ba7b-c15886e780a4.png)

tmcconechy commented on demos/ids-header/example.html:
```diff
@@ -49,3 +49,26 @@
+
+</br>
+
+<ids-text font-size="12" type="h1">Ids Header with Search Field:</ids-text>
```
In this example i think we should put it on its own page. What i would actually suggest is.

1) The search field one should be the new default ids-header/example.html
2) The previous one should be  ids-header/toolbar.html

vivz753 replied:
Done. Does ids-header not have an index.yaml for a reason? Should one be added? How do users know if there is a link to the /toolbar.html?

EdwardCoyle replied:
@vivz753 Probably not a bad idea to add one, I think Header was added before we standardized on the yaml format for listing examples

vivz753 replied:
Added an index.yaml

tmcconechy commented on demos/ids-search-field/example.html:
```diff
@@ -0,0 +1,57 @@
+<ids-layout-grid>
+  <ids-text font-size="12" type="h1">Search Field</ids-text>
```
In this example we have a lot of H1s i think we should just have this first one.

For the others can put it in the labels. Fx
```
<ids-search-field label="Products" value="Normal Search Field"></ids-search-field>
<ids-search-field label="Products" value="Disabled"></ids-search-field>
```

vivz753 replied:
Done

tmcconechy commented on demos/ids-trigger-field/example.html:
```diff
@@ -98,4 +98,44 @@
+    </ids-trigger-field>
+  </ids-layout-grid-cell>
+
+  <ids-layout-grid-cell>
```
This page now has 2 identical readonly search fields? Maybe redundant with the search page.

vivz753 replied:
Good catch. Removed duplicate

tmcconechy commented on src/components/ids-input/ids-input.scss:
```diff
@@ -71,6 +71,17 @@ $input-compact-height: $input-field-height-xs;
 .ids-input {
   @include block();
 
+  // ALTERNATE COLORS
```
```suggestion
  // alternate colors for header placement
```

tmcconechy commented on src/components/ids-search-field/TODO.md:
```diff
@@ -0,0 +1,27 @@
+- [x] add disabled
```
Could just leave whats left in terms of todos, also add title like the others

vivz753 replied:
Removed

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,251 @@
+ */
+
+const DEFAULT_LABEL = 'Search';
+const DEFAULT_PLACEHOLDER = 'Type to search';
```
Think maybe move this up so the comment is directly above the @customElement (with no spaces below the comment)

vivz753 replied:
Done

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,251 @@
+@customElement('ids-search-field')
+@scss(styles)
+class IdsSearchField extends mix(IdsElement).with(...appliedMixins) {
+  DEFAULT_LABEL = DEFAULT_LABEL;
```
Duplicate? See above?

vivz753 replied:
Added these because they weren't accessible otherwise through the IdsSearchField instance in the functional tests

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,251 @@
+      // TODO: pop up autocomplete suggestions
+    });
+
+    // this.onEvent('click', this.triggerButton, () => {
```
Generally we remove commented out code? Are all these TODO's still valid?

vivz753 replied:
Yeah, these are basically phase 2 features. Makes sense to leave them in there unless you have a strong reason not to

tmcconechy commented on src/components/ids-search-field/ids-search-field.js:
```diff
@@ -0,0 +1,251 @@
+    // });
+  }
+
+  #attachKeyboardListener() {
```
Make comments on each method/getter/setter ect

tmcconechy commented on src/components/ids-search-field/ids-search-field.scss:
```diff
@@ -0,0 +1,35 @@
+@import '../../core/ids-base';
+
+// Most of the css is added by ids-trigger field
+// Temporary fix until dropdown is merged TODO
```
Still need this?

vivz753 replied:
Removed

tmcconechy commented on src/components/ids-trigger-field/TODO.md:
```diff
@@ -0,0 +1,6 @@
+- [x] refactor state css selectors to be an attribute instead of a class
```
Add title remove old todo's?

vivz753 replied:
wdym add title?

vivz753 replied:
Removed old todo's

tmcconechy reviewed (changes requested):
Looks nice 👍🏻  A few nit picks and suggestions

tmcconechy reviewed (changes requested):
Looks nice 👍🏻  A few nit picks and suggestions

tmcconechy commented:
@vivz753 i can't tell if this is done so i added WIP label. Just take that off when you have no more pushes (also maybe try to do a bunch of commits and then push and not pus every time as it generates lots of builds and lots of emails)

vivz753 commented:
> @vivz753 i can't tell if this is done so i added WIP label. Just take that off when you have no more pushes (also maybe try to do a bunch of commits and then push and not push every time as it generates lots of builds and lots of emails)

I'll keep that in mind. It's done now

deep7102 reviewed (commented):
Looks Good! few things saw as

http://localhost:4300/ids-search-field
- focus styles
- colors on high-contrast theme

<img width="250" alt="image" src="https://user-images.githubusercontent.com/38439203/135632454-42c9f17f-346c-4c8d-a788-745303630fda.png">
<img width="334" alt="image" src="https://user-images.githubusercontent.com/38439203/135632705-ff2242d6-2bb1-475e-909e-66d9b9dc38db.png">

http://localhost:4300/ids-trigger-field

- since we added `Search Field` examples to trigger-field, should we change the icons to something else for other two example on this page which currently use magnifying glass (`Products`, and `Disabled Trigger Field`)
<img width="727" alt="image" src="https://user-images.githubusercontent.com/38439203/135633831-dbf52c11-2f65-42cf-a4b6-fda11af568a1.png">

vivz753 commented:
Adjusted the color variant for ids-input to show white text for new-contrast, and classic-light, classic-contrast. Also removed the search field examples from ids-trigger-field demo since ids-search-field exists now. @deep7102 Thanks for catching those

tmcconechy reviewed (approved):
Looks good to me. cc @deep7102

tmcconechy commented:
@vivz753 i spoke to deepak and now i am a bit confused. as he mentioned the focus state was off and i saw this issue https://github.com/infor-design/enterprise-wc/issues/376 which i dont understand.

Did you want to just make those fixes are is the idea to do that later on that issue?

deep7102 reviewed (approved):
Thanks! @vivz753 looks good

vivz753 commented:
> @vivz753 i spoke to deepak and now i am a bit confused. as he mentioned the focus state was off and i saw this issue #376 which i dont understand.
> 
> Did you want to just make those fixes are is the idea to do that later on that issue?

If I'm interpreting correctly, it's a styling issue for clearable focus, and out of scope for this component, so I made another issue.

tmcconechy commented:
Sure can be done later... will merge this for now
